Append content hash to `__git_commit`, if there are local uncommitted changes

vulkoingim opened the pull request:
## Description
<!-- Describe your changes in detail -->
* Append content hash to git commit, if there are local uncommitted changes
   * Traverses all dependencies and if there are changes to any of them appends the hash to them + the parents as well
* Adds the `__git_commit_short` arg that outputs the shortened version of the git commit to 7 chars

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/14467

## How to test
<!-- Provide steps to test this PR -->
See tests

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
* Append content hash to git commit, if there are local uncommitted changes
* Adds the `__git_commit_short` arg that outputs the shortened version of the git commit to 7 chars
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

meysholdt commented on pkg/leeway/package.go:
```diff
@@ -1002,7 +1006,65 @@ func resolveBuiltinGitVariables(pkg *Package, vars map[string]string) error {
+		}
+	}
+
+	for _, d := range pkg.GetTransitiveDependencies() {
```
Since `GetTransitiveDependencies()` returns not only all direct dependencies of this package but also all indirect dependencies, it's redundant to `isPkgDirty()` beeing recursive. Proposal:
Either call `GetTransitiveDependencies()` and don't recurse, or call `pkg.GetDependencies()` and keep the recursion.

meysholdt commented:
I left one comment - the other code changes LGTM.

vulkoingim commented:
> We should also update the Build arguments part of the README ([here](https://github.com/gitpod-io/leeway#build-arguments)) to include the new variable `__git_commit_short` as well as update the existing documentation to mention the new behaviour of appending the Leeway package version if the working copy is dirty ☺️

Modified the `README`

> Did you try to use this locally in a gitpod-io/gitpod repository together with `leeway run dev/preview:deploy-gitpod` to see if things work as expected?

Actually `deploy-gitpod` failed, but due to another reason 🤔  (not leeway related)
```bash
docker: Error response from daemon: manifest for eu.gcr.io/gitpod-core-dev/build/installer:aa-test-leeway-dev not found: manifest unknown: Failed to fetch "aa-test-leeway-dev" from request "/v2/gitpod-core-dev/build/installer/manifests/aa-test-leeway-dev".
```

Changing a file in a package and building it with leeway does produce the expected output:

```docker
./leeway build components/dashboard:docker -Dversion=test
...
[components/dashboard:docker] built image: eu.gcr.io/gitpod-core-dev/build/dashboard:test
[components/dashboard:docker] built image: eu.gcr.io/gitpod-core-dev/build/dashboard:commit-067fe9dae3f716b140673dc82b283f9c00979164-62b680cb4cd88a45b58cf098ea66ac1fb326c509
[components/dashboard:docker] package build succeded (60.34s)
☁️  uploading build artifacts to remote cache
```

Same for a package with a dependency (in this case `common-go` was changed. 


```docker
leeway build components/content-service:docker -Dversion=test
🔧  build  components/common-go:lib               (version 8a7a91c482e9772194ea17e9e79d726a1dc853c8)
🔧  build  components/content-service-api/go:lib  (version b31e25de6851db9d41e48d7e0a216d5ee2a3a4a2)
🔧  build  components/content-service:app         (version 9bbf039ce0a35cee27345c20367635ff2fdf124d)
🔧  build  components/content-service:docker      (version 1381f80f2d41ba046402a0f0c9ee7d1688974de0)

[components/common-go:lib] build started (version 8a7a91c482e9772194ea17e9e79d726a1dc853c8)
....
[components/content-service:docker] commit-067fe9dae3f716b140673dc82b283f9c00979164-1381f80f2d41ba046402a0f0c9ee7d1688974de0: digest: sha256:1c39eba39fbcf316c51289f6ee1d9e1acd690beb84f5d560c2d1b75232d564d6 size: 950
[components/content-service:docker] built image: eu.gcr.io/gitpod-core-dev/build/content-service:test
[components/content-service:docker] built image: eu.gcr.io/gitpod-core-dev/build/content-service:commit-067fe9dae3f716b140673dc82b283f9c00979164-1381f80f2d41ba046402a0f0c9ee7d1688974de0
[components/content-service:docker] package build succeded (20.63s)
```

> Finally, Leeway releases do use the CHANGELOG section of the PRs to produce releases (AFAIK) so would be nice to include a little changelog entry in this PR ☺️

Added a note as well :)
